PAY-10501: parse qr cpm

ap-sendy-agustian opened the pull request:
- add QrCpmConfiguration -> for @Bean cpm sub tag, app template sub tag, app specific transparent sub tag.
- add QrCpmParser -> main logic to parse from raw base64 to each object representation
- add decode base 64 and overload convertAlphaNumericHexToString
- add QrCpmPayload
- add unit test


example result taken from unit test 
     * DATA_MOCK_BASE64_3 all available tag = 85, 61 > 4F, 50, 5A, 5F20, 5F2D, 5F50, 63 >> 9F74 : 
<img width="1542" alt="Screenshot 2024-04-25 at 10 12 05" src="https://github.com/astrapay/spring-qris/assets/97210514/29b5a37a-5243-4444-b4ad-35671a65ce04">


<img width="1551" alt="Screenshot 2024-04-25 at 10 13 18" src="https://github.com/astrapay/spring-qris/assets/97210514/76fa082e-6947-4fe9-b31b-6e140f194302">

ap-asep-agus-heri committed:
PAY-25723 - Sebagai AstraPay, saya ingin dapat melakukan validasi atas qr text supaya dapat memastikan bahwa qr text sudah sesuai dengan format ASPI
- membuat validator mandatory field
- membuat validator conditional field
- membuat validator untuk hardcoded hexvalue
- membuat validasi hexlength

ap-asep-agus-heri committed:
PAY-25723 - Sebagai AstraPay, saya ingin dapat melakukan validasi atas qr text supaya dapat memastikan bahwa qr text sudah sesuai dengan format ASPI
- membuat validator mandatory field
- membuat validator conditional field
- membuat validator untuk hardcoded hexvalue
- membuat validasi hexlength
- membuat validasi untuk track 2 equivalent data

ap-asep-agus-heri committed:
Merge branch 'daily/sendy' of github.com:astrapay/spring-qris into daily/asep

ap-asep-agus-heri committed:
PAY-25723 - Sebagai AstraPay, saya ingin dapat melakukan validasi atas qr text supaya dapat memastikan bahwa qr text sudah sesuai dengan format ASPI
- membuat validator mandatory field
- membuat validator conditional field
- membuat validator untuk hardcoded hexvalue
- membuat validasi hexlength
- membuat validasi untuk track 2 equivalent data

ap-asep-agus-heri committed:
PAY-25723 - Sebagai AstraPay, saya ingin dapat melakukan validasi atas qr text supaya dapat memastikan bahwa qr text sudah sesuai dengan format ASPI
- membuat validator mandatory field
- membuat validator conditional field
- membuat validator untuk hardcoded hexvalue
- membuat validasi hexlength
- membuat validasi untuk track 2 equivalent data
- betulin unit test

ap-sendy-agustian committed:
Update QrCpmPayload.java

ap-sendy-agustian committed:
Update QrisHexConverter.java

ap-sendy-agustian committed:
add parse cpm

ap-sendy-agustian committed:
add unit test and configuration bean

ap-sendy-agustian committed:
Merge branch 'main' into daily/sendy

ap-sendy-agustian committed:
Update README.md

ap-sendy-agustian committed:
Update QrCpmParserTest.java

ap-sendy-agustian committed:
Update QrCpmParserTest.java

ap-sendy-agustian committed:
Update QrCpmParserTest.java

ap-sendy-agustian committed:
adjust unit test and cpm parser

ap-sendy-agustian committed:
Update QrCpmParserTest.java

ap-sendy-agustian committed:
Merge remote-tracking branch 'origin/daily/asep' into daily/testz

ap-sendy-agustian committed:
Update QrCpmPayload.java

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/QrCpmParser.java:
```diff
@@ -1,5 +1,153 @@
+        qrCpmPayload.setQrisRoot(qrisCpmMap);
+    }
+
+    private void parser(List<String> qrisCpmSubTag, String payloadHexBased, Map<String, QrCpmDataObject> qrisCpmMap) throws DecoderException {
```
Please add comments about what are this method do and why.
Can be added here or inside its scope (ahead of the while-scope and ahead of the if-scope).

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/QrCpmParser.java:
```diff
@@ -1,5 +1,153 @@
+        parser(qrisCpmSubTag, payloadHexBased, qrisCpmMap);
+    }
+
+    private void parseApplicationTemplate(Map<String, QrCpmDataObject> qrCpmDataObject) throws DecoderException {
```
Add context for this method too.

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/QrCpmParser.java:
```diff
@@ -1,5 +1,153 @@
+        }
+    }
+
+    private void parseApplicationSpecificTransparentTemplate(Map<String, QrCpmDataObject> qrCpmDataObject) throws DecoderException {
```
This one too.

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/object/QrCpmPayload.java:
```diff
@@ -1,5 +1,51 @@
+    private String payloadBase64;
+    private String payloadHex;
+
+    @CpmMandatoryField
```
Add link to the specification to help understanding these parities better.

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/validation/HexLengthValidator.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public boolean isValid(Map<String, QrisDataObject> value, ConstraintValidatorContext context) {
```
Add context to this method especially for what happens in line 38 for in the current version.

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/validation/Track2EquivalentDataValidator.java:
```diff
@@ -0,0 +1,77 @@
+            var hexString = value.get(this.id).getValue();
+            hexString = hexString.replaceAll("\\s+", "");
+            var fieldSeparatorIndex = hexString.indexOf(FIELD_SEPARATOR);
+            if (hexString.length() % EVEN_NUMBER_CHECKER == FIRST_INDEX) {
```
Prefer fast return instead of if-else whenever possible.

Example:
```java
            if (!FIRST_INDEX.equals(hexString.length() % EVEN_NUMBER_CHECKER)) {
                return false;
            }
            
            discretionaryData = hexString.substring(fieldSeparatorIndex + SERVICE_CODE_LAST_INDEX, hexString.length()); // Discretionary Data
```

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/validation/Track2EquivalentDataValidator.java:
```diff
@@ -0,0 +1,77 @@
+            if (pan.length() > PAN_MAX_LENGTH) {
+                return false;
+            }
+            ;
```
is this additional semicolon (;) a typo?

ap-asep-agus-heri replied:
typo

ap-yanuz-nurchalik commented on src/main/java/com/astrapay/qris/cpm/validation/Track2EquivalentDataValidator.java:
```diff
@@ -0,0 +1,77 @@
+            }
+            ;
+
+            result = fieldSeparatorString.equals(FIELD_SEPARATOR)
```
Can we just return instead of introducing another variable named result here?

e.g:
```java
            return fieldSeparatorString.equals(FIELD_SEPARATOR)
                    && isValidNibbleLength(fieldSeparatorString, FIELD_SEPARATOR_LENGTH)
                    && isValidNibbleLength(expirationDate, EXPIRATION_DATE_LENGTH)
                    && isValidNibbleLength(serviceCode, SERVICE_CODE_LENGTH);
```

ap-asep-agus-heri committed:
PAY-25723 - Sebagai AstraPay, saya ingin dapat melakukan validasi atas qr text supaya dapat memastikan bahwa qr text sudah sesuai dengan format ASPI
- membuat validator mandatory field
- membuat validator conditional field
- membuat validator untuk hardcoded hexvalue
- membuat validasi hexlength
- membuat validasi untuk track 2 equivalent data
- betulin unit test
- resolve comments

ap-sendy-agustian committed:
add context

ap-sendy-agustian committed:
Merge branch 'daily/sendy-cpm' of https://github.com/astrapay/spring-qris into daily/sendy-cpm
